### PR TITLE
spec: allow neon extension auto-upgrade + softfail upgrade

### PR DESCRIPTION
**Pull request by skyzh**
## Problem

reverts https://github.com/neondatabase/neon/pull/7128, unblocks https://github.com/neondatabase/cloud/issues/10742

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: compute_tools/src/spec.rs [hunk 1]
```diff
@@ -743,21 +743,21 @@ pub fn handle_extension_neon(client: &mut Client) -> Result<()> {
...
-    info!("handle neon extension upgrade (not really)");
-    // DISABLED due to compute node unpinning epic
-    // let query = "ALTER EXTENSION neon UPDATE";
-    // info!("update neon extension version with query: {}", query);
-    // client.simple_query(query)?;
+pub fn handle_neon_extension_upgrade(client: &mut Client) -> Result<()> {
+    info!("handle neon extension upgrade");
+    let query = "ALTER EXTENSION neon UPDATE";
+    info!("update neon extension version with query: {}", query);
+    client.simple_query(query)?;
```
Comment (skyzh) [inline compute_tools/src/spec.rs:760]: this is already soft-fail, the outer function handles the fail logic

**Review by save-buffer — COMMENTED** [review]
Is there a simple way to add a test to this? If it's not easy to do then maybe not worth it to add one, but just a thought

File: compute_tools/src/spec.rs [hunk 2]
```diff
@@ -743,21 +743,21 @@ pub fn handle_extension_neon(client: &mut Client) -> Result<()> {
...
     // - extension was just installed
     // - extension was already installed and is up to date
-    // DISABLED due to compute node unpinning epic
-    // let query = "ALTER EXTENSION neon UPDATE";
-    // info!("update neon extension version with query: {}", query);
-    // client.simple_query(query)?;
+    let query = "ALTER EXTENSION neon UPDATE";
+    info!("update neon extension version with query: {}", query);
+    if let Err(e) = client.simple_query(query) {
+        error!("failed to upgrade neon extension during `handle_extension_neon`: {e}");
```
[the remark this answers is not in the document]
Comment (skyzh) [reply compute_tools/src/spec.rs:749]: Oh I'll need to figure that out, `{e}` usually works for println! in newer version of Rust, but not sure if it works for `error!`.

**Comment by skyzh** [comment]
> Is there a simple way to add a test to this? If it's not easy to do then maybe not worth it to add one, but just a thought

Could be part of the compatibility test, I can give a try.

File: compute_tools/src/spec.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (save-buffer) [reply compute_tools/src/spec.rs:749]: Ok, sounds good, I just hadn't seen it before but if you've seen it work then it should be fine :)

**Comment by skyzh** [comment]
added test case in https://github.com/neondatabase/neon/pull/7231/commits/e6cbf6b336311214fb2dccf29775926640363bcb

**Review by save-buffer — APPROVED** [review]
Nice!

**Pull request merged**